salt/states/x509_v2.py

```diff
@@ -1578,8 +1579,24 @@ def _file_managed(name, test=None, **kwargs):
         raise SaltInvocationError("test param can only be None or True")
     # work around https://github.com/saltstack/salt/issues/62590
     test = test or __opts__["test"]
-    res = __salt__["state.single"]("file.managed", name, test=test, **kwargs)
-    return res[next(iter(res))]
+    file_managed = __states__["file.managed"]
+    if test:
+        # calls via __salt__["state.single"](..., test=test)
```

Comment:
Shouldn't the fix be in `state.single` if that call is what is causing the overwrite in `__opts__`?

Reply:
I suspect the problem is somewhere in the loader, though I'm neither qualified nor motivated to find out tbh. I have tried to debug a very similar problem with overriding `__opts__` not being reversed in the Vault runner before. Also not sure if this workaround can be applied to `state.single` easily.

This specific bug breaks the x509_v2 state module pretty hard though and I suspect it will take a bit of time until someone figures out the root cause, if at all (since it's not a huge deal for most of Salt).

Reply:
@Ch3LL can we merge it as is? It's a huge issue in 3006.x since there is a problem installing mcrypto to onedir and the only other way to use x509 state is to use x509_v2 which has that issue.

Reply:
@lkubb Mind checking the `file.directory` issue reported here https://github.com/saltstack/salt/issues/64195#issuecomment-1613547558 ?

Reply:
@Oloremo Without having much to go on regarding the report, I fail to see how `file.directory` would behave differently with this patch atm. If anything, I would expect it the other way around (`test` to not be respected). Using `file.directory` instead of `file.managed` for the reproduction test in this PR succeeds as well. So my current assumption is that it's an unrelated issue or the patch was not applied correctly, barring any future (and more detailed) reports.

Reply:
Confirmed that the minion cache has loaded the patched module.
Some additional details, which might add to things not caught by the test

- There's a csr_managed
- There's a private_key_managed
- There's a file.directory
- The csr_managed has a require requisite on the private_key_managed, which in turn has a require requisite on the file.directory
- There is a unrelated file.managed in the same state file

Reply:
Going to have @dwoz take a look at this one, as he is more familiar with the loader.

Reply:
> This bug is nicely distilled in this issue #62590.

Do you mean that bug has the same source as the sticky `__opts__` causing this issue with `state.single` or fixing that bug will remove the necessity for using `state.single` instead of `__states__` here in the first place?

The problem is that the `x509_v2` state module also overrides `test` for early checks to avoid signing certificates that are discarded directly after, e.g. by a missing parent dir without `makedirs`. So if it does not have the same cause, a restructuring of the `x509_v2` state module would still be necessary (not the end of the world though).

Reply:
> > This bug is nicely distilled in this issue #62590.
> 
>  fixing that bug will remove the necessity for using `state.single` instead of `__states__` here in the first place?
> 
I don't think we should need to refactor `x509_v2` after fixing the bug.

Reply:
@dwoz 
It's been a month, any updates on this issue? I believe it block literally everyone who wants to use 3006.x and x509

Reply:
#64960 addresses this issue. I think the PR can be closed.